deepspeed/runtime/zero/stage3.py

```diff
@@ -520,11 +556,25 @@ def _all_gather(self, partitioned_params, async_op=False):
 
     def _synchronize_communication(self, synchronize_streams=True):
         assert len(self.params_in_flight) == len(self.in_flight_handles)
-        for handle, param in zip(self.in_flight_handles, self.params_in_flight):
-            if handle is not None:
-                with torch.cuda.stream(self.comm_stream):
-                    handle.wait()
-            param.ds_status = ZeroParamStatus.AVAILABLE
+
+        if len(self.params_in_flight) > 0:
+            avail_params = [param.ds_id for param in self.params_in_flight]
+            status_params = [param.ds_status for param in self.params_in_flight]
+
+            with torch.cuda.stream(self.comm_stream):

```

Comment:
The `with torch.cuda.stream(self.comm_stream)` statement is not going to take effect on the underlying *communication* operations, as far as i know. The reason is that the `torch.distributed.all_gather` uses its own cuda stream. 

Code reference point:
https://github.com/pytorch/pytorch/blob/v1.9.0/torch/lib/c10d/ProcessGroupNCCL.cpp#L1111

Reply:
@zarzen, thanks. I will try this. Also, I see some great progress in your PR #1188. Can you see our we can merge both efforts?

Reply:
@zarzen, that sounds good. I was not suggesting merging into a single PR. What I meant was to avoid conflicts and duplications in both efforts. For now, we don't need to do anything but just be aware of the relationship of both efforts. Thanks.